### PR TITLE
Introducing the Picostates architecture

### PR DESCRIPTION
**Pull request by cowboyd**
## Purpose

Microstates has been rewritten six times in the past in the past two years. While the vision has remained consistent: to provide a way to make immutable changes to composable objects, the implementation has had to evolved with our understanding of the problem domain, and also with the evolution of our requirements. Each time, we’ve pushed the boundaries of what has been possible for us given our knowledge at the time, and each time we’ve come away with something more powerful. Yet each time, as we reach a new level of capability, we encounter problems as we try to take on the next level of issues that the new implementation has revealed.

In this instance, the introduction of stable states meant that we gained the feature of stability, but at the expense that the tree we were using to represent the microstates became very complex. The question then, is how to simplify it while also accounting for the next key feature: stable identities for microstates?

The current iteration of the microstates architecture represents an abstract tree as its primary source of truth. In representing the parent child relationship of the various microstates, it contains the path of each node, and the overall context, but the actual objects that the user sees that contains the transitions are metadata of the nodes of this tree. The following diagram shows this relationship.

Given this tree as the primary source of truth:

picostates

The microstate objects (containing the transitions), and the `state` object containing the raw value are computed off the properties in this tree.

picostates

The prolem with this arrangement is that the individual micostate nodes are _not_ connected by JavaScript reference, and so before we can present them to user space, we have to go through the process of “collapsing” them into a set of normal references. In other words, walking the tree of nodes, and creating a version of that node's microstates that references the microstates of that nodes children. 

picostates

This means that the nodes on the tree are not the actual objects that will be presented to the user and so making those user-land objects stable presents something of a challenge. While it’s achievable, it increases the complexity both in the code and also in the computation, and makes working with other things like middleware and queries more difficult.

 ## Approach

The fundamental shift in architecture introduced by this PR flips this relationship on its head. The microstate tree, and the state tree are _always the primary source of truth_. When a transition happens, this tree is modified in place, and you can work with it directly. The metadata nodes containing information about the object’s path in the tree are maintained as a set of _detached metadata_ where each piece of metadata only contains information about the individual microstate. This is actually more convenient because the metadata does not need to be connected to other metadata in the tree, and as a result the amount of metadata that needs to be stored is a lot less.

This diagram indicates the news scheme.

picostates

As you can see, the Microstate objects reference their children directly, and it also referernces the parallel tree of state directly, all of which are there all the time. In this conception, it is the metadata that is hanging out, uncollapsed from the microstate nodes.

This has the result of making transitions very, very simple because it’s nothing more than a lens set, and we lean on our lens for what it means to set one microstate into another.

Let's look at the lens `set` operation:

```js
set(lens, value, rootContext);
```

It takes 3 arguments. The lens itself which indicates where in the tree we are focused (the path). The actual value that we want to set at that path, and the root context which is the whole entire tree. 

Under the picostates architecture, we take 2 out of these three arguments (lens, and root context) and stuff them in the metadata of microstate. Then, for every microstate `set(value)` we can extract the implicit lens, and the implicit root context from the metadata to convert `set(value)` -> `set(lens, value, rootContext)`

Of coures all of this hinges on our own implementation of our own `SubstateAt` lens which keeps the metadata, the microstate, and the state tree all in sync as we add each successifve layer of context. But because we only have to deal with one layer of context at a time, we don’t have to deal with operations that run up and down the tree at the same time. We only have to solve for a single case of adding one layer of context, and then all the other cases of incrementally adding layer after layer of context are taken care of for us.

### Identity

Finally, identity (stable proxy) is provide for the microstate tree by implementing a special Tree functor that maps one native JavaScript tree to another. This was a key innovation.: you can map any object, not just an explicit tree.  

Using a microstate tree comprised of native JavaScript references, we can create an exactly structured JavaScript object graph except with the type and conten of objects swapped out at each point. This lets us build a proxy tree out of a microstate tree with only a few lines of code. 

### Breaking Changes

There are a few breaking changes with the move to this architecture. While they are breaking, they are actually a move towards a much more intuitive architecture.

1. `create(Type, value)` always returns an instance of `Type`. This was actually something counter-intuitive (and also a headache for potential typescript users). 
1. `microstate.state` is always just a pojo corresponding to the `value` argument to create. Previously, it was an instance of `Type`
1. `microstate.valueOf()` is not special anymore, and will just perform the normal JavaScript `valueOf()` (which will just return the microstate itself). In cases where you wanted to use `valueOf()` you would now just use the `state` property.
1. Transitions are _no longer allowed_ inside of a computed property.

### Todos

- [x] Rollup builds are not currently sorted out
- [ ] identities do not currently map queries properly
- [ ] verification for laziness and recursive microstates.

### CODE REVIEW

**Comment by taras** [comment]
@cowboyd this pr is missing `Microstates.from`

File: tests/examples/cart.test.js [hunk 1]
```diff
@@ -1,14 +1,18 @@
...
+import { reduce, sum } from '../../src/query';
 
 describe('cart example', () => {
   class Cart {
-    products = Array;
+    products = create(ArrayType);
+
     get price() {
-      return this.products.reduce((acc, product) => acc + product.quantity * product.price, 0);
+      return reduce(this.products, (acc, product) => acc + product.state.quantity * product.state.price, 0);
```
[the remark this answers is not in the document]
Comment (taras) [reply tests/examples/cart.test.js]: These are great questions.

> Is there a reason you would want to have access to the microstates wrapper object in one of these selector/query function definitions?

`this` in the context of a query is the local microstate. It's the node at that particular location. It's not the wrapper object. 

> These functions should be side effect free, so you wouldn't be calling a transition in here.

Yes, these are not transitions. They're pure functions. They just know how to operate on microstates.

This branch of microstates has a [new typeclass called Reducible](https://github.com/microstates/microstates.js/blob/picostates/src/query.js#L4) which describes how a microstate of a type can be reduced. You can see an implementation of this in [Array's Reducible instance](https://github.com/microstates/microstates.js/blob/picostates/src/types/array.js#L54). 

These describe how an instance of any microstate can be reused and these operations are pure. 

> If there isn't a reason, then could we have the query functions (reduce, sum, etc) unwrap the state POJO (e.g. product.state) before calling the predicate or accumulator function?

If we do this, we will not be able to return microstates from getters. That will mean that we won't be able to invoke transitions from getters. It'll be very limiting to Microstates.

These are really good questions. There is a lot to document still. 

Please keep the questions coming.

[the remark this answers is not in the document]
Comment (rkeeler) [reply tests/examples/cart.test.js]: > If we do this, we will not be able to return microstates from getters. That will mean that we won't be able to invoke transitions from getters. It'll be very limiting to Microstates.

That makes sense. In this particular case where the reduce function isn't returning a microstate but is just calculating a raw value (e.g. the price of the shopping cart), that fact was lost on me. But if you had an array of microstates that you were filtering to be a smaller array with a subset of those microstates, you would want to be able to call transitions on any of those values in the result of the getter.

Thanks for the explanation!

**Comment by cowboyd** [comment]
> this pr is missing Microstates.from

@taras Do you think this should block this PR? It's a really cool feature, and definitely doable here, but given that there is a lot of breaking changes to the API, would you say that this is on the critical path?

**Comment by cowboyd** [comment]
> If there isn't a reason, then could we have the query functions (reduce, sum, etc) unwrap the state POJO (e.g. product.state) before calling the predicate or accumulator function?

I concur with everything @taras said here and would add the following items.

1. While transitions should not be invoked within a getter, you may want to access _other_ computed properties within the getter, or within the reduction, so you'd need the whole state.
2. It's so much conceptually simpler to always be operating on microstates and never have to context-switch. One of the biggest stumbling blocks we encountered with adoption on prior versions of microstates was confusion about when the context was a microstate, when it was a state, or when it was a value.

Your intuition is on-point though that there is definitely a fuzzy area and why we only operated on state in prior versions.

**Comment by cowboyd** [comment]
Note: Query mapping and lazy verification will happen as follow-on pull requests.

**Pull request merged**